src/xtgeo/grid3d/grid.py

```diff
@@ -297,6 +304,141 @@ def grid_from_surfaces(
+
+        self.top_depth = surface_from_grid3d(
+            one, where="top", property="depth", rfactor=4, index_position="top"
+        )
+        self.base_depth = surface_from_grid3d(
+            one, where="base", property="depth", rfactor=4, index_position="base"
+        )
+        self.top_i_index = surface_from_grid3d(
+            one, where="top", property="i", rfactor=4, index_position="top"
+        )
+        self.top_j_index = surface_from_grid3d(
+            one, where="top", property="j", rfactor=4, index_position="top"
+        )
+        self.base_i_index = surface_from_grid3d(
+            one, where="base", property="i", rfactor=4, index_position="base"
+        )
+        self.base_j_index = surface_from_grid3d(
+            one, where="base", property="j", rfactor=4, index_position="base"
+        )
+
+        # need to fill (interpolate) eventual holes in maps
+        self.top_depth.fill()
+        self.base_depth.fill()
+        self.top_i_index.fill()
+        self.top_j_index.fill()
+        self.base_i_index.fill()
+        self.base_j_index.fill()
+
+        self.onegrid_cpp = new_grid_cpp
+        self.grid_cpp = grid_cpp
+        self.top_depth_cpp = _internal.regsurf.RegularSurface(self.top_depth)
+        self.base_depth_cpp = _internal.regsurf.RegularSurface(self.base_depth)
+        self.top_i_index_cpp = _internal.regsurf.RegularSurface(self.top_i_index)
+        self.top_j_index_cpp = _internal.regsurf.RegularSurface(self.top_j_index)
+        self.base_i_index_cpp = _internal.regsurf.RegularSurface(self.base_i_index)
+        self.base_j_index_cpp = _internal.regsurf.RegularSurface(self.base_j_index)
+
+        # these are special SWIG array objects kept until necessary SWIG
+        # methods are replaced
+        self.top_i_index_carr = self._get_swig_carr_double(self.top_i_index)
+        self.top_j_index_carr = self._get_swig_carr_double(self.top_j_index)
+        self.base_i_index_carr = self._get_swig_carr_double(self.base_i_index)
+        self.base_j_index_carr = self._get_swig_carr_double(self.base_j_index)
+
+        logger.info("Initialized cache for grid %s", grid.name)
+
+    def clear(self) -> None:
+        """Clear the cache."""
+        self.onegrid = None
+        self.bbox = None
+        self.top_depth = None
+        self.base_depth = None
+        self.top_i_index = None
+        self.top_j_index = None
+        self.base_i_index = None
+        self.base_j_index = None
+
+        self.onegrid_cpp = None
+        self.top_depth_cpp = None
+        self.base_depth_cpp = None
+        self.top_i_index_cpp = None
+        self.top_j_index_cpp = None
+        self.base_i_index_cpp = None
+        self.base_j_index_cpp = None
+
+        self.top_i_index_carr = None
+        self.top_j_index_carr = None
+        self.base_i_index_carr = None
+        self.base_j_index_carr = None
+
+        logger.info("Clear cache for grid %s", self.name)
+
+
```

Comment:
The grid cache Class replaces the earlier self._tmp construct, although there are still places where this is used. The cache significantly speed ups some computations.